temporalcli/commands.workflow_reset.go

```diff
@@ -0,0 +1,205 @@
+
+	resp, err := cl.ResetWorkflowExecution(cctx, &workflowservice.ResetWorkflowExecutionRequest{
+		Namespace: c.Parent.Namespace,
+		WorkflowExecution: &common.WorkflowExecution{
+			WorkflowId: c.WorkflowId,
+			RunId:      resetBaseRunID,
+		},
+		Reason:                    fmt.Sprintf("%s: %s", username(), c.Reason),
+		WorkflowTaskFinishEventId: eventID,
+		ResetReapplyType:          reapplyType,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to reset workflow: %w", err)
+	}
+
+	if cctx.JSONOutput {
+		return cctx.Printer.PrintStructured(
+			resp,
+			printer.StructuredOptions{})
+	}
+	return nil
+}
+
+func (c *TemporalWorkflowResetCommand) getResetEventIDByType(ctx context.Context, cl client.Client) (string, int64, error) {
+	resetType, namespace, wid, rid := c.Type.Value, c.Parent.Namespace, c.WorkflowId, c.RunId
+	wfsvc := cl.WorkflowService()
+	switch resetType {
+	case "LastWorkflowTask":
+		return getLastWorkflowTaskEventID(ctx, namespace, wid, rid, wfsvc)
+	case "LastContinuedAsNew":
+		return getLastContinueAsNewID(ctx, namespace, wid, rid, wfsvc)
+	case "FirstWorkflowTask":
+		return getFirstWorkflowTaskEventID(ctx, namespace, wid, rid, wfsvc)
+	default:
+		return "", -1, fmt.Errorf("invalid reset type: %s", resetType)
+	}
+}
+
+// Returns event id of the last completed task or id of the next event after scheduled task.
+func getLastWorkflowTaskEventID(ctx context.Context, namespace, wid, rid string, wfsvc workflowservice.WorkflowServiceClient) (resetBaseRunID string, workflowTaskEventID int64, err error) {
+	resetBaseRunID = rid
+	req := workflowservice.GetWorkflowExecutionHistoryReverseRequest{
+		Namespace: namespace,
+		Execution: &common.WorkflowExecution{
+			WorkflowId: wid,
+			RunId:      rid,
+		},
+		MaximumPageSize: 250,
+		NextPageToken:   nil,
+	}
+
+	for more := true; more; more = len(req.NextPageToken) != 0 {
+		resp, err := wfsvc.GetWorkflowExecutionHistoryReverse(ctx, &req)
+		if err != nil {
+			return "", 0, fmt.Errorf("failed to get workflow execution history: %w", err)
+		}
+		for _, e := range resp.GetHistory().GetEvents() {
+			if e.GetEventType() == enums.EVENT_TYPE_WORKFLOW_TASK_COMPLETED {
+				workflowTaskEventID = e.GetEventId()
+				break
+			} else if e.GetEventType() == enums.EVENT_TYPE_WORKFLOW_TASK_SCHEDULED {
+				// if there is no task completed event, set it to first scheduled event + 1
+				workflowTaskEventID = e.GetEventId() + 1
+			}
+		}
+		req.NextPageToken = resp.NextPageToken
+	}
+	if workflowTaskEventID == 0 {
+		return "", 0, errors.New("unable to find any scheduled or completed task")
+	}
+	return
+}
+
+// Returns id of the first workflow task completed event or if it doesn't exist then id of the event after task scheduled event.
+func getFirstWorkflowTaskEventID(ctx context.Context, namespace, wid, rid string, wfsvc workflowservice.WorkflowServiceClient) (resetBaseRunID string, workflowTaskEventID int64, err error) {
+	resetBaseRunID = rid
```

Comment:
I do not see you ever change `resetBaseRunID` or `rid` in the function.

Reply:
Yep. All three variants (these are lightly-cleaned up version of the original logic) use the same API even though only `getLastContinueAsNewID` changes it. We could change two of the three calls in `getResetEventIDByType` to do, say, `return rid, getLastWorkflowTaskEventID(...)` but I didn't